core/iwasm/interpreter/wasm_runtime.c

```diff
@@ -2078,6 +2078,57 @@ wasm_instantiate(WASMModule *module, bool is_sub_inst,
             &module_inst->e->functions[module->retain_function];
     }
 
+    if (exec_env_main
+        && ((WASMModuleInstance const *)(exec_env_main->module_inst))
+                   ->e->c_api_func_imports
+               != NULL) {
+        /* workaround about passing instantiate-linking information */
+        {
```

Comment:
@cpetig It is to fix the issue of `c_api_func_imports are not passed to new threads` reported in #2149 , right? My suggestion is to apply the changes to `lib_wasi_threads_wrapper.c` like #2170 since (1) it is only related to multi-threading, adding code here may increase the footprint, (2) the fixing for AOT mode is invalid, the related code is in `aot_instantiate`: https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/aot/aot_runtime.c#L1089

@eloparco How about we merging the fixing of #2170 to this PR? Just use this PR to fix all these issues? Thanks.

Reply:
BTW, there may be duplicated code in `lib_wasi_threads_wrapper.c` and `lib_pthread_wrapper.c`, after discussing with @lum1n0us, we suggest to add an API in `core/iwasm/libraries/thread-mgr/thread_manager.{h|c}`, e.g.,
```C
bool
wasm_cluster_dup_c_api_imports(WASMModuleInstance *module_inst_dst,
                               const WASMModuleInstance *module_inst_src);
```
@cpetig @eloparco How about that?

Reply:
*lib_pthread_wrapper.c* has a same workaround. It doesn't seem clever to write the same workaround twice. Shall we merge them together and store it somewhere public, maybe *thread_manager*?

Reply:
> How about we merging the fixing of https://github.com/bytecodealliance/wasm-micro-runtime/pull/2170 to this PR? Just use this PR to fix all these issues? Thanks.

Makes sense, I'll close the other PR.

> lib_pthread_wrapper.c has a same workaround. It doesn't seem clever to write the same workaround twice. Shall we merge them together and store it somewhere public, maybe thread_manager?

Yes, that's what I had in mind as well when in my PR I wrote "refactor common code".

Reply:
> Do I understand you correctly that https://github.com/cpetig/wasm-micro-runtime/blob/main/core/iwasm/libraries/thread-mgr/thread_manager.c#L758 is not necessary because this is taken care of in aot_runtime.c ?

I think it's probably the other way around, at the beginning you were putting your changes in `wasm_runtime.c` so you would have needed to change `aot_runtime.c` as well. If the changes are in the pthread and wasi thread files no change in `aot_runtime.c` should be required.